modules/host/persist.go

```diff
@@ -1,80 +1,165 @@
+func (h *Host) loadObligations(cos []contractObligation) {
+	// Clear the existing obligations maps.
+	for i := range cos {
+		obligation := &cos[i] // both maps should use same pointer
 		height := obligation.FileContract.WindowStart + StorageProofReorgDepth
+		// Sanity check - if the height is below the current height, then set
+		// the height to current height + 3. This makes sure that all file
+		// contracts will eventually be hit or garbage collected by the host,
+		// even if a bug means that they aren't acted upon at the right moment.
+		if build.DEBUG && height < h.blockHeight {
+			panic("host settings file is inconsistent")
+		} else if height < h.blockHeight {
+			height = h.blockHeight + 3
+		}
 		h.obligationsByHeight[height] = append(h.obligationsByHeight[height], obligation)
 		h.obligationsByID[obligation.ID] = obligation
 
-		// update spaceRemaining
-		if len(obligation.LastRevisionTxn.FileContractRevisions) > 0 { // COMPATv0.4.8
+		// update spaceRemaining to account for the storage held by this
+		// obligation.
+		if len(obligation.LastRevisionTxn.FileContractRevisions) > 0 {
 			h.spaceRemaining -= int64(obligation.LastRevisionTxn.FileContractRevisions[0].NewFileSize)
 		}
 	}
-	h.secretKey = sHost.SecretKey
-	h.publicKey = sHost.PublicKey
+}
+
+// load extrats the save data from disk and populates the host.
+func (h *Host) load() error {
+	p := new(persistence)
+	err := persist.LoadFile(persistMetadata, p, filepath.Join(h.persistDir, "settings.json"))
+	if err != nil {
+		return err
+	}
+
+	// Copy over the host context.
+	h.blockHeight = p.BlockHeight
+	h.recentChange = p.RecentChange
+	h.publicKey = p.PublicKey
+	h.secretKey = p.SecretKey
+	h.settings = p.Settings
+
+	// Copy over the file management. The space remaining is recalculated from
+	// disk instead of being saved, to maximize the potential usefulness of
+	// restarting Sia as a means of eliminating unkonwn errors.
+	h.fileCounter = p.FileCounter
+	h.spaceRemaining = p.Settings.TotalStorage
+	h.loadObligations(p.Obligations)
+
+	// Copy over statistics.
+	h.profit = p.Profit
 
 	return nil
 }
 
+// establishDefaults configures the default settings for the host, overwriting
+// any existing settings.
+func (h *Host) establishDefaults() error {
+	// Configure the settings object.
+	h.settings = modules.HostSettings{
+		TotalStorage: 10e9,     // 10 GB
+		MaxFilesize:  100e9,    // 100 GB - deprecated field
+		MaxDuration:  144 * 60, // 60 days
+		WindowSize:   288,      // 48 hours
+		Price:        defaultPrice,
+		Collateral:   types.NewCurrency64(0),
+	}
+	h.spaceRemaining = h.settings.TotalStorage
+
+	// Generate signing key, for revising contracts.
+	sk, pk, err := crypto.GenerateKeyPair()
+	if err != nil {
+		return err
+	}
+	h.secretKey = sk
+	h.publicKey = types.SiaPublicKey{
+		Algorithm: types.SignatureEd25519,
+		Key:       pk[:],
+	}
+
+	// Save the defaults to disk.
+	err = h.save()
+	if err != nil {
+		return err
+	}
+	return nil
+}
```

Comment:
Maybe it's a difference of top-down vs. bottom-up organization. I didn't like how complicated the `New` function was, so I split it into parts that initialized the persistence (logger + settings), and parts that initialized the network components.

Maybe there's a better organization, but I think that splitting it up is better than keeping it all in the primary function. I think it makes sense to call `establishDefaults` after loading fails, as opposed to previously where the defaults were established and then overwritten by the `load` call (which I believe actually was causing a bug where one value wasn't properly being loaded, so instead was always the default).

Network stuff on the other hand doesn't have any persistence, and should be called every time.

Or maybe the host struct itself should be organized by how the fields get initialized. It's pretty clear that all of the module dependencies are set in `New`, but the defaults are split across Host context and file management, and stuff like the height is pretty confusing too.